neqo-bin/src/server/http09.rs

```diff
@@ -127,67 +106,79 @@ impl HttpServer {
             }
             return;
         }
+        let read_buffer = &self.read_buffer[..sz];
 
-        data.truncate(sz);
-        let buf = if let Some(mut existing) = self.read_state.remove(&stream_id) {
-            existing.append(&mut data);
-            existing
-        } else {
-            data
-        };
+        let buf = self.read_state.remove(&stream_id).map_or(
+            Cow::Borrowed(read_buffer),
+            |mut existing| {
+                existing.extend_from_slice(read_buffer);
+                Cow::Owned(existing)
+            },
+        );
 
         let Ok(msg) = std::str::from_utf8(&buf[..]) else {
-            self.save_partial(stream_id, buf, conn);
+            self.save_partial(stream_id, buf.to_vec(), conn);
             return;
         };
 
         let m = self.regex.captures(msg);
         let Some(path) = m.and_then(|m| m.get(1)) else {
-            self.save_partial(stream_id, buf, conn);
+            self.save_partial(stream_id, buf.to_vec(), conn);
             return;
         };
 
-        let resp = {
+        let resp: ResponseData = {
             let path = path.as_str();
             qdebug!("Path = '{path}'");
             if self.is_qns_test {
                 match qns_read_response(path) {
-                    Ok(data) => Some(data),
+                    Ok(data) => data.into(),
                     Err(e) => {
                         qerror!("Failed to read {path}: {e}");
-                        Some(b"404".to_vec())
+                        b"404".to_vec().into()
                     }
                 }
             } else {
                 let count = path.parse().unwrap();
-                Some(vec![b'a'; count])
+                ResponseData::zeroes(count)
```

Comment:
Previously when requesting say a 10 GiB response from an `http09.rs` server, the server would pre-allocate the full 10 GiB. With this commit it will allocate `STREAM_IO_BUFFER_SIZE` (32k) up-front only and cycle through the buffer until the full response is sent.